core/src/main/java/com/linecorp/armeria/common/stream/ConcatPublisherStreamMessage.java

```diff
@@ -0,0 +1,280 @@
+            }
+        }
+
+        void nextSource() {
+            upstream.request(1);
+        }
+    }
+
+    private static final class InnerSubscriber<T> extends SubscriptionArbiter implements Subscriber<T> {
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<InnerSubscriber> cancelledUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(InnerSubscriber.class, "cancelled");
+
+        private Subscriber<? super T> downstream;
+        private final SubscriptionOption[] options;
+
+        @Nullable
+        private OuterSubscriber<T> outerSubscriber;
+
+        private volatile boolean completed = true;
+        // Updated via cancelledUpdater
+        private volatile int cancelled;
+
+        InnerSubscriber(Subscriber<? super T> downstream, SubscriptionOption[] options) {
+            this.downstream = downstream;
+            this.options = options;
+        }
+
+        /**
+         * Sets the {@link OuterSubscriber}.
+         * This should be set before subscribing to upstream.
+         */
+        void setOuterSubscriber(OuterSubscriber<T> outerSubscriber) {
+            this.outerSubscriber = outerSubscriber;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (isCancelled()) {
+                subscription.cancel();
+                return;
+            }
+            // Reset 'completed' to subscribe to new Publisher
+            completed = false;
+            setUpstreamSubscription(subscription);
+            outerSubscriber.inSubscribe = false;
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+            if (isCancelled()) {
+                return;
+            }
+            downstream.onNext(item);
+            produced(1);
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            downstream.onError(throwable);
+        }
+
+        @Override
+        public void onComplete() {
+            if (isCancelled()) {
+                return;
+            }
+            completed = true;
+            if (outerSubscriber.completed) {
+                downstream.onComplete();
+            } else {
+                outerSubscriber.nextSource();
+            }
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                downstream.onError(new IllegalArgumentException(
+                        "Rule §3.9 violated: non-positive requests are forbidden"));
+                return;
+            }
+            if (isCancelled()) {
+                return;
+            }
+            super.request(n);
+        }
+
+        @Override
+        public void cancel() {
+            if (cancelledUpdater.compareAndSet(this, 0, 1)) {
+                if (outerSubscriber != null && outerSubscriber.upstream != null) {
```

Comment:
Could you make another `if` and extract `outerSubscriber.upstream` so that we don't access the volatile field twice?